File: client/client/app/util/imm_api.go
package util

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

var SERVER_URL string
var v = GetViper()

func init() {
	SERVER_URL = v.GetString("server_url")
}

func get_token(username string, password string) (string, error) {
	// get token from config
	token := v.GetString("IMM_TOKEN")
	Logger.Debug("token: ", token)
	if token != "" {
		// check if it's expired
		expire := v.GetInt("IMM_EXPIRE")
		Logger.Debug("expire: ", expire)
		if expire != 0 {
			if time.Now().Unix() < int64(expire) {
				return token, nil
			} else {
				Logger.Info("Token expired, refreshing...")
			}
		}
	} else {
		fmt.Println("No token saved, retrieving...")
	}
	return refresh_token(username, password)
}

func refresh_token(username string, password string) (string, error) {
	data := url.Values{}
	data.Set("username", username)
	data.Set("password", password)

	resp, err := http.PostForm(SERVER_URL+"/"+"login", data)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("invalid response status code: %d", resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var tokenData struct {
		AccessToken string `json:"access_token"`
		ExpiresIn   int64  `json:"expires_in"`
	}
	err = json.Unmarshal(body, &tokenData)
	if err != nil {
		return "", err
	}

	// token expires in 9 hours
	expire := time.Now().Add(9 * time.Hour).Unix()

	v.Set("IMM_TOKEN", tokenData.AccessToken)
	v.Set("IMM_EXPIRE", strconv.FormatInt(expire, 10))

	// Save the changes to the config file
	err = v.WriteConfig()
	if err != nil {
		fmt.Println(fmt.Errorf("failed to write config file: %w", err))
	}

	return tokenData.AccessToken, nil
}

func gen_request_headers() (http.Header, error) {
	token, err := get_token(v.GetString("imm_account"), v.GetString("imm_password"))
	if err != nil {
		return nil, err
	}
	headers := http.Header{}
	headers.Add("accept", "application/json")
	headers.Add("Content-Type", "application/json")
	headers.Add("Authorization", "Bearer "+token)
	return headers, nil
}

func Imm_api_post(endpoint string, reqBody []byte) (*http.Response, error) {
	url := SERVER_URL + "/" + endpoint
	Logger.Debug("imm_api_post enpoint url: ", url)

	headers, err := gen_request_headers()
	if err != nil {
		return &http.Response{}, err
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(reqBody))
	if err != nil {
		Logger.Error(err)
	}
	req.Header = headers

	// Send request
	resp, err := http.DefaultClient.Do(req)

	// request error
	if err != nil {
		return &http.Response{}, fmt.Errorf("imm_api_post request error: %v", err)
	}

	// response error
	if resp.StatusCode != http.StatusOK {
		return &http.Response{}, fmt.Errorf("invalid response status code: %d", resp.StatusCode)
	}

	return resp, nil
}

func Imm_api_get(endpoint string, queries url.Values) (*http.Response, error) {
	url := SERVER_URL + "/" + endpoint
	Logger.Debug("imm_api_get enpoint url: ", url)

	headers, err := gen_request_headers()
	if err != nil {
		return &http.Response{}, err
	}

	req, err := http.NewRequest("GET1", url, nil)
	if err != nil {
		return &http.Response{}, err
	}
	req.Header = headers

	// Send request
	resp, err := http.DefaultClient.Do(req)

	// request error
	if err != nil {
		return &http.Response{}, err
	}

	// response error
	if resp.StatusCode != http.StatusOK {
		return &http.Response{}, fmt.Errorf("invalid response status code: %d", resp.StatusCode)
	}

	return resp, nil

}

// delete and put is not tested yet TODO:
func Imm_api_delete(endpoint string) (*http.Response, error) {
	url := SERVER_URL + "/" + endpoint

	headers, err := gen_request_headers()
	if err != nil {
		return &http.Response{}, err
	}

	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		fmt.Println(err)
	}
	req.Header = headers

	// Send request
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Println(err)
	}

	return resp, nil
}

func Imm_api_put(endpoint string, queries url.Values) (*http.Response, error) {
	url := SERVER_URL + "/" + endpoint

	headers, err := gen_request_headers()
	if err != nil {
		return &http.Response{}, err
	}

	req, err := http.NewRequest("PUT", url, nil)
	if err != nil {
		fmt.Println(err)
	}
	req.Header = headers

	// Send request
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Println(err)
	}

	return resp, nil
}

File: client/system/imm_api.py
import token
import requests
import dotenv
import os
from datetime import datetime
import base64
import json
from client.system.config import SERVER_URL, config, config_path

AUTH_URL = SERVER_URL + "login"
print(f"We are using server {SERVER_URL}")


def get_token(username: str, password: str):
    # get token from env
    token = config.get("imm_token")
    if token is not None:
        # check if it's expired
        expire = config.get("imm_expire")
        if expire is not None and datetime.now().timestamp() < int(expire):
            return token
        else:
            print("Token expired, refreshing...")
    else:
        print("No token saved, retrieving...")
    return refresh_token(username, password)


def refresh_token(username: str, password: str):
    data = {"username": username, "password": password}
    try:
        resp = requests.post(url=AUTH_URL, data=data)
        resp.raise_for_status()

        token = resp.json().get("access_token")
        # JWT token = header + '.' + payload + '.' + signature
        jwt_segs = token.split(".")
        encoded_payload = jwt_segs[1]
        # decode with correct padding
        decoded_payload = base64.b64decode(
            encoded_payload + "=" * (-len(encoded_payload) % 4)
        )
        expire = json.loads(decoded_payload).get("exp")
        # update env file to keep latest token and expire values
        with open(config_path, "w") as f:
            config["imm_token"] = token
            config["imm_expire"] = str(expire)
            json.dump(config, f)

        return token
    except requests.exceptions.HTTPError as errh:
        print(errh)
        if resp.status_code == 401:
            print("Incorrect username/password.")
        return
    except requests.exceptions.ConnectionError as errc:
        print(errc)
        return


def gen_request_headers():
    imm_account = config.get("imm_account")
    imm_password = config.get("imm_password")
    token = get_token(imm_account, imm_password)

    headers = {
        "accept": "application/json",
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}",
    }
    return headers


def imm_api_post(endpoint: str, data=None):
    resp = requests.post(
        SERVER_URL + endpoint, json=data, headers=gen_request_headers()
    )
    return resp


def imm_api_get(endpoint: str, queries=None):
    resp = requests.get(
        SERVER_URL + endpoint, params=queries, headers=gen_request_headers()
    )
    return resp


def imm_api_delete(endpoint: str):
    resp = requests.delete(SERVER_URL + endpoint)
    return resp


def imm_api_put(endpoint: str, params=None):
    resp = requests.put(SERVER_URL + endpoint, params=params)
    return resp
